File: backend/src/main/java/de/prime_ux/backend/cases/Case.java
package de.prime_ux.backend.cases;

import de.prime_ux.backend.tenants.Tenant;
import de.prime_ux.backend.triage.CaseCategory;
import de.prime_ux.backend.triage.CaseTier;
import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.EnumType;
import jakarta.persistence.Enumerated;
import jakarta.persistence.FetchType;
import jakarta.persistence.Id;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.ManyToOne;
import jakarta.persistence.Table;
import java.math.BigDecimal;
import java.time.Instant;
import java.util.UUID;
import lombok.AccessLevel;
import lombok.Getter;
import lombok.NoArgsConstructor;
import org.hibernate.annotations.UuidGenerator;

/**
 * A case ("Vorgang") — the central entity of frontdesk. Every ingested mail becomes a case and
 * later travels through triage, drafting, and approval. The case holds the facts about the mail
 * that opened it and the verdict on it; the mails themselves — the opening one, what followed,
 * what went out — are its {@link CaseMessage}s.
 */
@Entity
@Table(name = "cases")
@Getter
@NoArgsConstructor(access = AccessLevel.PROTECTED)
public class Case {

	@Id
	@UuidGenerator
	private UUID id;

	@ManyToOne(fetch = FetchType.LAZY, optional = false)
	@JoinColumn(name = "tenant_id")
	private Tenant tenant;

	@Column(name = "message_id")
	private String messageId;

	@Column(nullable = false)
	private String sender;

	// The tenant address the mail was sent to (info@, rechnung@, ...); null when
	// the mail carries no usable recipient header.
	@Column
	private String recipient;

	@Column(nullable = false)
	private String subject;

	// When the opening mail came in. The dashboard counts arrivals by it.
	@Column(name = "received_at", nullable = false)
	private Instant receivedAt;

	// When the conversation last moved, in either direction: the list sorts by it.
	@Column(name = "last_message_at", nullable = false)
	private Instant lastMessageAt;

	@Column(name = "ingested_at", nullable = false)
	private Instant ingestedAt;

	@Column(name = "has_attachments", nullable = false)
	private boolean hasAttachments;

	@Column(name = "size_bytes", nullable = false)
	private long sizeBytes;

	// The triage's verdict; null until it ran, which is how the runner finds the
	// cases still waiting for it.
	@ManyToOne(fetch = FetchType.LAZY)
	@JoinColumn(name = "category_id")
	private CaseCategory category;

	@Enumerated(EnumType.STRING)
	@Column
	private CaseTier tier;

	// The model's own certainty, kept so the threshold can be tuned against
	// reality rather than guessed.
	@Column
	private BigDecimal confidence;

	@Column(name = "triaged_at")
	private Instant triagedAt;

	// When a person took note of the case. Null while it is still waiting to be
	// looked at, which is how the review finds what is left to work through.
	@Column(name = "handled_at")
	private Instant handledAt;

	// When somebody threw the case away. It then sits in the trash rather than
	// being gone: a mail cannot be fetched again once the mailbox marked it read.
	@Column(name = "deleted_at")
	private Instant deletedAt;

	// The model's one-sentence answer to "what does the sender want?".
	@Column
	private String summary;

	// The reply as the model wrote it, kept so it can be measured later how much
	// people had to change. Null when nobody asked the model, or a person wrote
	// the draft from scratch.
	@Column(name = "draft_generated_text")
	private String draftGeneratedText;

	// The next reply as it stands: what a person edits, and what the send button
	// sends. Null means the case has no draft, which is how the runner finds the
	// cases still waiting for one; cleared again once the reply went out.
	@Column(name = "draft_text")
	private String draftText;

	@Column(name = "draft_generated_at")
	private Instant draftGeneratedAt;

	@Column(name = "draft_updated_at")
	private Instant draftUpdatedAt;

	/** A case as its opening mail makes it; the mail itself goes in as the first message. */
	public Case(Tenant tenant, String messageId, String sender, String recipient, String subject, Instant receivedAt,
			boolean hasAttachments, long sizeBytes) {
		this.tenant = tenant;
		this.messageId = messageId;
		this.sender = sender;
		this.recipient = recipient;
		this.subject = subject;
		this.receivedAt = receivedAt;
		this.lastMessageAt = receivedAt;
		this.ingestedAt = Instant.now();
		this.hasAttachments = hasAttachments;
		this.sizeBytes = sizeBytes;
	}

	/**
	 * The customer wrote again. Whatever was settled is open again: the case goes back into the
	 * inbox, and the paperclip stays on once anything in the conversation had an attachment.
	 * Category, tier and a reply somebody started to write are left as they are. Refused in the
	 * trash — what was thrown away does not come back through the customer.
	 */
	public void receiveFollowUp(Instant receivedAt, boolean withAttachments) {
		requireNotTrashed();
		this.handledAt = null;
		this.lastMessageAt = receivedAt;
		this.hasAttachments = this.hasAttachments || withAttachments;
	}

	/**
	 * A person overruling the triage. Only the tier moves: the category, the summary and the
	 * confidence stay what the model said, because they still describe the classification it made
	 * — what changed is what happens with the case, and that is the tenant's call.
	 */
	public void changeTier(CaseTier tier) {
		this.tier = tier;
	}

	/**
	 * A person filing the case under another category, or under none at all. The tier stays where
	 * it is: which category a mail belongs to and what happens with it are two decisions, and a
	 * correction to the one is not a correction to the other.
	 */
	public void changeCategory(CaseCategory category) {
		this.category = category;
	}

	/**
	 * A person taking note of a case, or taking that back. The moment stands where a flag would
	 * do, because "since when" is what gets asked as soon as two people work the same inbox.
	 *
	 * <p>Marking an already handled case again leaves the first moment where it is: it is still
	 * the same taking note, and the second click says nothing new.
	 */
	public void markHandled(boolean handled) {
		if (!handled) {
			this.handledAt = null;
		}
		else if (this.handledAt == null) {
			this.handledAt = Instant.now();
		}
	}

	/**
	 * Thrown away, or fetched back out of the trash. The row stays either way — what leaves for
	 * good leaves through the repository, and only from the trash.
	 *
	 * <p>Throwing away says nothing about whether the case was worked through: it keeps its
	 * handled moment, so a case fetched back lands where it was, in the inbox or in the archive.
	 */
	public void moveToTrash() {
		if (this.deletedAt == null) {
			this.deletedAt = Instant.now();
		}
	}

	public void restore() {
		this.deletedAt = null;
	}

	/**
	 * Records what the triage made of this case. The category may be null when the model found
	 * nothing that fits; the tier never is, because every case has to land somewhere.
	 */
	public void applyTriage(CaseCategory category, CaseTier tier, BigDecimal confidence, String summary) {
		this.category = category;
		this.tier = tier;
		this.confidence = confidence;
		this.summary = summary;
		this.triagedAt = Instant.now();
	}

	/** Whether there is a reply to read: the current text, however it came to be. */
	public boolean hasDraft() {
		return this.draftText != null;
	}

	/**
	 * What the model wrote, replacing whatever draft there was — edits included. Both texts start
	 * out the same; they part ways with the first edit.
	 *
	 * <p>Refused for a case in the trash: nobody answers a mail that was thrown away, and the
	 * runner never asks for one, so this only guards the button.
	 */
	public void applyDraft(String text) {
		requireNotTrashed();
		Instant now = Instant.now();
		this.draftGeneratedText = text;
		this.draftText = text;
		this.draftGeneratedAt = now;
		this.draftUpdatedAt = now;
	}

	/**
	 * A person's version of the reply. What the model wrote stays where it is — that is the point
	 * of keeping it — and a person may write a draft where the model never did.
	 */
	public void editDraft(String text) {
		requireNotTrashed();
		this.draftText = text;
		this.draftUpdatedAt = Instant.now();
	}

	/**
	 * The reply went out and is a message of the conversation now, so the box is emptied for the
	 * next one. Sending is also taking note: the case leaves the inbox for the archive, unless
	 * somebody ticked it off before.
	 */
	public void markSent(Instant sentAt) {
		requireNotTrashed();
		this.draftText = null;
		this.draftGeneratedText = null;
		this.draftGeneratedAt = null;
		this.draftUpdatedAt = null;
		this.lastMessageAt = sentAt;
		if (this.handledAt == null) {
			this.handledAt = sentAt;
		}
	}

	private void requireNotTrashed() {
		if (this.deletedAt != null) {
			throw new IllegalStateException("A case in the trash gets no draft");
		}
	}
}

File: frontend/src/app/domains/cases/model/case.ts
/**
 * A case ("Vorgang") — one matter on its way through the pipeline, brought in by the mailbox or
 * written down by hand.
 * Mirrors the backend's CaseResponse, except that receivedAt arrives as an ISO
 * string and is parsed into a Date by the CasesService — the table's date
 * filter compares real Date objects.
 */
/**
 * How a case reached the house. Everything but `mail` was written down by hand and carries no
 * address to answer to — what stands in its sender is a person, not a mailbox.
 */
export type CaseChannel = 'mail' | 'phone' | 'fax' | 'other';

/** The channels a case can be written down under; mail comes from the mailbox, never from a form. */
export const MANUAL_CHANNELS = ['phone', 'fax', 'other'] as const satisfies readonly CaseChannel[];

export type ManualChannel = (typeof MANUAL_CHANNELS)[number];

/**
 * What the form hands over to write a case down. Lives here rather than with the service because
 * the dialog is a ui component, and ui may see the model and nothing else.
 */
export type NewCase = {
  channel: ManualChannel;
  contact: string;
  subject: string;
  text: string;
  categoryId: string | null;
  tier: CaseTier | null;
  /** Who is to handle it, if whoever took the call already knows. */
  assigneeId: string | null;
};

export type Case = {
  id: string;
  channel: CaseChannel;
  /** Who it came from: an address for a mail, whatever the person typed for anything else. */
  sender: string;
  /** The tenant address the mail was sent to; null for mails ingested before it was recorded. */
  recipient: string | null;
  subject: string;
  receivedAt: Date;
  hasAttachments: boolean;
  sizeBytes: number;
  /** What the triage made of the case; all of it null until it looked at the case. */
  summary: string | null;
  /** Which category the case sits in; a picker needs the key, the name is what is read. */
  categoryId: string | null;
  categoryName: string | null;
  /** The colour the category carries, if any; the row is drawn in it. */
  categoryColor: CaseCategoryColor | null;
  tier: CaseTier | null;
  /** The model's own certainty between 0 and 1 — a self-assessment, not a measured probability. */
  confidence: number | null;
  /**
   * When a person took note of the case. Null while it is still waiting to be looked at, which is
   * what the review works through. Deliberately not a deletion: the case stays in the inbox.
   */
  handledAt: Date | null;
  /**
   * When somebody threw the case away. It then sits in the trash, out of the inbox and out of the
   * archive, until it is deleted for good or fetched back.
   */
  deletedAt: Date | null;
  /** Whether a reply is waiting to be read. The text itself is the detail's. */
  hasDraft: boolean;
  /** Who has the case, or null while nobody has taken it; a picker needs the key, the list the name. */
  assigneeId: string | null;
  assigneeName: string | null;
  /** How many internal notes the case carries; what they say is the detail's. */
  noteCount: number;
  /** When the conversation last moved, in either direction; the list sorts and groups by it. */
  lastMessageAt: Date;
  /** How long the conversation is; more than one says the customer or the house wrote again. */
  messageCount: number;
};

/**
 * One case with everything the detail view shows. The body is deliberately not part of {@link
 * Case}: the list would pay for the full text of every mail on every reload and never shows one.
 * The draft flag goes the other way: the detail carries the draft itself, so the flag would say
 * nothing the text does not.
 */
export type CaseDetail = Omit<Case, 'hasDraft' | 'messageCount'> & {
  /**
   * The next reply as it stands: what the model wrote, or what a person made of it. Null while
   * the case has none, and again once a reply went out. The two moments say when the model wrote
   * it and when it last changed; a draft a person wrote from scratch has no first moment.
   */
  draftText: string | null;
  draftGeneratedAt: Date | null;
  draftUpdatedAt: Date | null;
  /** The conversation, oldest first: the mail that opened the case, what followed, what went out. */
  messages: CaseMessage[];
  /** What the case has been through, oldest step first. */
  events: CaseEvent[];
};

/**
 * One message of a case's conversation. An incoming one is a mail the customer sent, with its
 * text and, where it was written so, its HTML; an outgoing one is a reply that went out, text
 * only, in the name of whoever pressed the button. What came attached hangs off the message it
 * came with — metadata only, the bytes have an endpoint of their own.
 */
export type CaseMessage = {
  id: string;
  direction: 'incoming' | 'outgoing';
  sender: string;
  recipient: string | null;
  subject: string;
  bodyText: string;
  bodyHtml: string | null;
  /** Received, or sent. */
  occurredAt: Date;
  sizeBytes: number;
  sentByName: string | null;
  attachments: CaseAttachment[];
};

/** What can happen to a case, as the trail writes it down. */
export type CaseEventType =
  | 'ingested'
  | 'follow_up_received'
  | 'triaged'
  | 'classification_corrected'
  | 'draft_generated'
  | 'draft_edited'
  | 'note_deleted'
  | 'assigned'
  | 'unassigned'
  | 'handled'
  | 'reopened'
  | 'trashed'
  | 'restored'
  | 'sent';

/**
 * One step of a case's trail: what happened, when, who did it — null for what the system did on
 * its own — and a few facts about it whose keys depend on the type: the tier and confidence of a
 * verdict, the address a reply went to.
 */
export type CaseEvent = {
  type: CaseEventType;
  occurredAt: Date;
  actorName: string | null;
  details: Record<string, unknown>;
};

/**
 * One part that was attached to the mail. Inline parts are the pictures the HTML body shows in
 * place, a signature's logo mostly; they are put back into the body by their Content-ID and kept
 * out of the list of things a person would open.
 */
export type CaseAttachment = {
  id: string;
  fileName: string;
  /** The MIME type without parameters, e.g. `application/pdf`. */
  contentType: string;
  sizeBytes: number;
  inline: boolean;
  contentId: string | null;
};

/** A category as a case is filed under it: what it is called, and the colour it is drawn in. */
export type SelectableCategory = {
  id: string;
  name: string;
  color: CaseCategoryColor | null;
};

/**
 * An internal note on a case: what colleagues tell each other about it, and never the customer.
 * `updatedAt` is null while the note stands as it was written.
 */
export type CaseNote = {
  id: string;
  authorName: string;
  text: string;
  createdAt: Date;
  updatedAt: Date | null;
  /** Whether the person reading wrote it — the server decides, the page only offers accordingly. */
  own: boolean;
};

/** A colleague a case can be handed to: what the picker needs and nothing else. */
export type AssignableUser = {
  id: string;
  name: string;
};

/** What the assignee filter offers for a case nobody has taken; no user can carry this id. */
export const NOBODY = 'none';

/**
 * The palette a category's colour can come from. Spelled out here rather than imported from the
 * admin domain: bounded contexts keep their own model, and the inbox only ever reads the name to
 * resolve `--app-category-<name>` from styles.css.
 */
export type CaseCategoryColor = 'blue' | 'green' | 'amber' | 'red' | 'violet' | 'teal' | 'grey';

/**
 * How a case is handled: two questions resolved into one ladder — does this need an answer, and
 * who writes it. `info` needs none but should be seen, `ignore` needs neither.
 */
export type CaseTier = 'automatic' | 'draft' | 'manual' | 'info' | 'ignore';

/**
 * The three piles a case can be in, and the three pages that show them: what is left to do, what
 * was worked through, and what somebody threw away. Every case is in exactly one of them.
 */
export type CasePile = 'inbox' | 'archive' | 'trash';
